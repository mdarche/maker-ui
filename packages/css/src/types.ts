import type { Interpolation } from '@emotion/react'

export type Breakpoints = (string | number)[]
export type StyleObject = object | Interpolation<any>

/**
 * Support for the breakpoints prop and responsive CSS scales.
 *
 * @param breakpoints - An array of breakpoints that determines the css prop media queries
 * @param css - A css object that can support array values for all nested properties
 *
 * @remarks
 * Used by all Maker UI components, primitives, and custom jsx pragma.
 *
 */
export interface MakerProps {
  css?: StyleObject
  breakpoints?: Breakpoints
}

/**
 * A string, number, or array of either lets you stack values for media queries.
 *
 * @example A simple example:
 * <div css={{ height: [50, 100]}} />
 */
export type ResponsiveScale = string | number | (string | number)[]

declare module 'react' {
  interface Attributes extends MakerProps {}
}

/**
 * Direct export from EmotionJS
 */
export { Interpolation }
