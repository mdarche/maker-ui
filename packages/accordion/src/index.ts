export { Accordion, type AccordionProps } from './Accordion'
export { AccordionPanel, type AccordionPanelProps } from './AccordionPanel'
