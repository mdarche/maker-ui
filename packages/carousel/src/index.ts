export { Carousel, CarouselProps } from './Carousel'
