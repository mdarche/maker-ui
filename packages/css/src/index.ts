export { type GlobalProps, Global } from './Global'
export { jsx } from './jsx'

export * from './types'
