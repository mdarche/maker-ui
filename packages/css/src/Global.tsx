import * as React from 'react'
import { Global as EmotionGlobal } from '@emotion/react'

import { formatCSS } from './css'
import type { Interpolation } from './types'

export interface GlobalProps {
  styles: Interpolation<any> | object
  breakpoints?: (string | number)[]
}

/**
 * A global style declaration that supports responsive arrays and the `breakpoints` prop
 *
 * @param styles - a CSS style object
 * @param breakpoints - an array of breakpoints
 *
 * @link https://maker-ui.com/docs/elements/global
 */
export const Global = ({ styles, breakpoints }: GlobalProps) => {
  return (
    <EmotionGlobal
      styles={
        formatCSS(
          styles as Interpolation<any>,
          breakpoints
        ) as Interpolation<any>
      }
    />
  )
}
