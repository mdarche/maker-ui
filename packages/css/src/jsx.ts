import { createElement } from 'react'
import { jsx as emotionJsx, Interpolation } from '@emotion/react'

import { formatCSS } from './css'

/**
 * Scans JSX props for `breakpoints` and `css` to format responsive arrays
 *
 * @param props - all of the element's props
 * @returns All original props + a formatted, responsive EmotionJS `css` prop
 *
 * @internal usage only
 *
 */

function parseProps(props: any) {
  if (!props || !props.css) return props
  let next: typeof props & { css: Interpolation<any> } = {}

  for (let key in props) {
    if (key === 'css' || key === 'breakpoints') continue
    next[key] = props[key]
  }

  next.css = formatCSS(props.css, props.breakpoints)
  return next
}

/**
 * JSX that supports responsive arrays and the `breakpoints` prop. Uses normal
 * React.createElement if `breakpoints` and `css` are undefined..
 *
 * @remarks
 * This is just a prop-formatting wrapper for Emotion's jsx export
 *
 * @link https://maker-ui.com/docs/jsx
 *
 */

export const jsx = <P extends {}>(
  type: React.ElementType<P>,
  props: React.Attributes & P,
  ...children: React.ReactNode[]
) =>
  typeof props?.css !== 'undefined' || typeof props?.breakpoints !== 'undefined'
    ? emotionJsx(type, parseProps(props), ...children)
    : createElement(type, props, ...children)
