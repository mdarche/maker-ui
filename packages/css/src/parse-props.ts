import { formatCSS } from './css'
import type { Interpolation } from './types'

/**
 * Scans JSX props for `breakpoints` and `css` to format responsive arrays
 *
 * @param props - all of the element's props
 * @returns All original props + a formatted, responsive EmotionJS `css` prop
 *
 * @internal usage only
 *
 */
export function parseProps(props: any) {
  if (!props || !props.css) return props
  let next: typeof props & { css: Interpolation<any> } = {}

  for (let key in props) {
    if (key === 'css' || key === 'breakpoints') continue
    next[key] = props[key]
  }

  next.css = formatCSS(props.css, props.breakpoints)
  return next
}
