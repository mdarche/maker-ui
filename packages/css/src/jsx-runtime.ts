import { createElement } from 'react'
import { jsx as emotionJsx } from '@emotion/react'
import { parseProps } from './parse-props'
import type { MakerProps, StyleObject } from './types'

declare module 'react' {
  interface Attributes extends MakerProps {}
}

/**
 * Scans the CSS prop and returns true if every key is undefined
 *
 * @param {StyleObject} obj a CSS prop style object
 * @returns {boolean}
 */
function isEmpty(obj: StyleObject) {
  return Object.values(obj as object).every(
    (el) => el === undefined || el === null
  )
}

/**
 * JSX that supports responsive arrays and the `breakpoints` prop. Uses normal
 * React.createElement if `breakpoints` and `css` are undefined..
 *
 * @remarks
 * This is just a prop-formatting wrapper for Emotion's jsx export
 *
 * @link https://maker-ui.com/docs/jsx
 *
 */
export const jsx = <P>(
  type: React.ElementType<P>,
  props: React.Attributes & P,
  ...children: React.ReactNode[]
) => {
  if (props?.css && isEmpty(props.css)) {
    delete props.css
    delete props?.breakpoints
  }

  return props?.css
    ? emotionJsx(type as string, parseProps(props), ...children)
    : createElement(type as string, props, ...children)
}
