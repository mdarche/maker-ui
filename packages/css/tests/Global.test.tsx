import * as React from 'react'
import '@testing-library/jest-dom'
import { render, screen } from '@testing-library/react'
import { Global } from '../src/Global'

const color = 'rgb(251, 251, 251)'
const breakpoints = [450, 768]

describe('Global style component', () => {
  test('global styles are added to the DOM', () => {
    render(<Global styles={{ body: { background: color } }} />)

    const body = document.querySelector('body')
    const style = body ? window.getComputedStyle(body).background : undefined
    expect(style).toEqual(color)
  })

  test('global parses nested objects and responsive properties', () => {
    render(
      <>
        <Global
          breakpoints={breakpoints}
          styles={{
            h2: { fontSize: [24, 28] },
            a: { textDecoration: 'none' },
          }}
        />
        <h2 id="heading">Heading</h2>
        <a href="/">Link</a>
      </>
    )

    expect(screen.getByText('Link')).toHaveStyle({ textDecoration: 'none' })
    expect(screen.getByText('Heading')).toHaveStyle({ fontSize: '24px' })

    // Test for media queries in snapshot
    expect(document.head).toMatchSnapshot()
  })
})
