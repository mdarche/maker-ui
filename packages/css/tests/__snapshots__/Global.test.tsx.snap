// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Global style component global parses nested objects and responsive properties 1`] = `
<head>
  <style
    data-emotion="css-global"
    data-s=""
  >
    
    h2{font-size:24px;}
  </style>
  <style
    data-emotion="css-global"
    data-s=""
  >
    
    @media screen and (min-width: 450px){h2{font-size:28px;}}
  </style>
  <style
    data-emotion="css-global"
    data-s=""
  >
    
    a{-webkit-text-decoration:none;text-decoration:none;}
  </style>
</head>
`;

exports[`global correctly parses nested objects and responsive properties 1`] = `
<head>
  <style
    data-emotion="css-global"
    data-s=""
  >
    
    h2{font-size:24px;}
  </style>
  <style
    data-emotion="css-global"
    data-s=""
  >
    
    @media screen and (min-width: 450px){h2{font-size:28px;}}
  </style>
  <style
    data-emotion="css-global"
    data-s=""
  >
    
    a{-webkit-text-decoration:none;text-decoration:none;}
  </style>
</head>
`;
